keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,659 @@
+    - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+    - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposal aims at enabling hot plugging of node compute resources. This will help in updating cluster resource capacity by just resizing compute resources of nodes rather than adding new node to a cluster.
+The updated node configurations are to be reflected at the node and cluster levels automatically without the need to reset the kubelet.
+
+This proposal also aims to improve the initialization and reinitialization of resource managers, such as the CPU manager and memory manager, in response to changes in a node's CPU and memory configurations.
+
+## Motivation
+In a typical Kubernetes environment, the cluster resources may need to be altered due to following reasons:
+- Incorrect resource assignment during cluster creation.
+- Increased workload over time, leading to the need for additional resources in the cluster.
+
+To handle these scenarios, we can:
+- Horizontally scale up the cluster by adding compute nodes.
+- Vertically scale up the cluster by increasing node capacity. However, currently, the workaround for capturing node resizing in the cluster involves restarting the Kubelet.
+
+Node resource hot plugging will provide advantages in scenarios such as:
+- Handling resource demand with a limited set of nodes by increasing the capacity of existing nodes instead of creating new nodes.
+- Creating new nodes takes more time compared to increasing the capacity of existing nodes.
+
+### Goals
+
+* Dynamically scale up the node by hot plugging  resources and without restarting the kubelet.
+* Ability to reinitialize resource managers (CPU manager, memory manager) to adopt changes in node's resource.
+
+### Non-Goals
+
+* Dynamically adjust system reserved and kube reserved values.
+* Hot unplug of node resources.
```

Comment:
I am not convinced it is right to capture hot unplug in the non-goal section. I understand that it is currently not in scope for alpha due to certain challenges but those should be well documented in the KEP for posterity and considering we might want to pursue scale down at later stage of this KEP e.g. Beta or GA.

Reply:
The reason behind having Hot unplug of node resources in the Non-Goals is to have propose separate KEP in future based on the community feedback for the scale down/unplug of resource. Hence to we renamed this KEP from Node Dynamic resize to Node Resource Hot plug with more focus towards upscaling the resoureces.

Reply:
when we break something down to pieces like this, we try to outline the path forward. Any time we just vaguely postpone the hard problem for the future, we end up not merging solution for a simple problem as some tradeoffs in code are harder to make without knowing how harder problems will be addressed

Reply:
breaking it into two was my idea as it seemed more accessible to do hot plug because then kubelet doesn't have to go through admission again. would you prefer the two features be tackled at once @SergeyKanzhelev ? or just the downsize case be though through while we're considering the upsize

Reply:
@SergeyKanzhelev has a point here. The breakdown also makes sense to me, but we need to make sure there's an incremental path forward and that, when we tackle hotunplug, we can be reasonnably sure we can build on these foundations and not have to undo some changes.

I think the middleground could be start outlining the hard parts with hotunplug and start sketching out the needs we have to solve them.

Reply:
Sure, Currently I have added the hot unplug under the scope for Future Work section in this KEP however please let use know we need to add in more details to this same KEP or if it is ideal to start working on another KEP regarding this in parallel.

Reply:
details are required in this KEP. It is not a good way to slice the feature as accounting for the node downsize may have implications for the upsize

Reply:
Added a section about Handling HotUnplug Events with some details.

Reply:
There were concerns regarding HotUnplug so we wanted to clearly mention it as non-goal for this KEP, 
However it was also necessary to highlight sequence events incase of HotUnplug. we only wanted to support hotunplug as informing event rather than functionally handling it in a future KEP.

Reply:
Sounds like hot-unplug, even though listed as future work, is completely decoupled from this KEP. 

Naively thinking, the way hot-plug is designed/implemented is likely to affect how hot-unplug can be done the future. Could we talk about why the current design will be compatible with hot-unplug even with the known challenges there?

Reply:
closing the loop here: https://github.com/kubernetes/enhancements/pull/5585